paimon-core/src/main/java/org/apache/paimon/rest/RESTCatalog.java

```diff
@@ -0,0 +1,199 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.rest;
+
+import org.apache.paimon.annotation.VisibleForTesting;
+import org.apache.paimon.catalog.Catalog;
+import org.apache.paimon.catalog.Database;
+import org.apache.paimon.catalog.Identifier;
+import org.apache.paimon.fs.FileIO;
+import org.apache.paimon.fs.Path;
+import org.apache.paimon.manifest.PartitionEntry;
+import org.apache.paimon.options.Options;
+import org.apache.paimon.rest.requests.ConfigRequest;
+import org.apache.paimon.rest.responses.ConfigResponse;
+import org.apache.paimon.schema.Schema;
+import org.apache.paimon.schema.SchemaChange;
+import org.apache.paimon.table.Table;
+
+import org.apache.paimon.shade.jackson2.com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+/** A catalog implementation for REST. */
+public class RESTCatalog implements Catalog {
+    private RESTClient client;
+    private String token;
+    private ResourcePaths resourcePaths;
+    private Map<String, String> options;
+    private Map<String, String> baseHeader;
+
+    private static final ObjectMapper objectMapper = RESTObjectMapper.create();
+
+    public RESTCatalog(Options options) {
+        String uri = options.get(RESTCatalogOptions.URI);
+        token = options.get(RESTCatalogOptions.TOKEN);
+        Optional<Duration> connectTimeout =
+                options.getOptional(RESTCatalogOptions.CONNECTION_TIMEOUT);
+        Optional<Duration> readTimeout = options.getOptional(RESTCatalogOptions.READ_TIMEOUT);
+        Integer threadPoolSize = options.get(RESTCatalogOptions.THREAD_POOL_SIZE);
+        int queueSize = options.get(RESTCatalogOptions.THREAD_POOL_QUEUE_SIZE);
+        HttpClientOptions httpClientOptions =
+                new HttpClientOptions(
+                        uri,
+                        connectTimeout,
+                        readTimeout,
+                        objectMapper,
+                        threadPoolSize,
+                        queueSize,
+                        DefaultErrorHandler.getInstance());
+        this.client = new HttpClient(httpClientOptions);
+        Map<String, String> initHeaders =
+                RESTUtil.merge(configHeaders(options.toMap()), authHeaders(token));
+        this.options = optionsInner(initHeaders, options.toMap());
+        this.baseHeader = configHeaders(this.options());
```

Comment:
We need to merge authHeader too?

Reply:
When set up, we need to merge the auth header. Later, we will support using header() to refresh the token.